BUILD.bazel

```diff
@@ -1,51 +1,81 @@
-    "//conditions:default": ":opentracing_unix",
-  }),
+
+hdrs = glob(["include/opentracing/**/*.h"]) + [
+    ":include/opentracing/config.h",
+    ":include/opentracing/version.h",
+]
+
+includes = ["include"]
+
+deps = [
+  "//3rd_party:expected",
+  "//3rd_party:variant",
+]
+
+## WINDOWS ##
+
+cc_windows_binary(
+  # must include .dll extension in a shared library name on Windows
+  name = "opentracing.dll",
+  srcs = glob(["src/**/*.cpp"], exclude = ["src/dynamic_load_unsupported.cpp", "src/dynamic_load_unix.cpp"]) + hdrs,
+  # when this binary is included by other programs, it will be dynamically linked
+  linkshared = 1,
+  includes = includes,
+  deps = deps,
+  # let the preprocessor know to include special Windows dll export symbols
+  defines = ["OPENTRACING_EXPORTS"],
+)
+
+# filegroup is a windows-specific quirk
+filegroup(
+  name = "opentracing_import_lib",
+  srcs = [":opentracing.dll"],
+  output_group = "interface_library",
+)
+
+cc_import(
+  name = "opentracing_dll_import",
+  interface_library = ":opentracing_import_lib",
+  shared_library = ":opentracing.dll",
+)
+
+## POSIX ##
+
+cc_posix_binary(
+  # must include .so extension in a shared library name for posix
+  name = "libopentracing.so",
+  srcs = glob(["src/**/*.cpp"], exclude = ["src/dynamic_load_unsupported.cpp", "src/dynamic_load_windows.cpp"]) + hdrs,
+  linkshared = 1,
+  deps = deps,
 )
 
-windows_dll_library(
-  name="opentracing_windows",
-  srcs = glob(["src/**/*.cpp"], exclude = ["src/dynamic_load_unsupported.cpp", "src/dynamic_load_unix.cpp"]),
-  hdrs = glob(["include/opentracing/**/*.h"]) + [
-      ":include/opentracing/config.h",
-      ":include/opentracing/version.h",
-  ],
-  # so that source files can write opentracing/... instead of include/opentracing/...
-  includes=["include"],
-  deps = [
-    "//3rd_party:expected",
-    "//3rd_party:variant",
-  ],
-  # sets the OPENTRACING_EXPORTS precompiler variable while building DLL
-  # so that header "import symbols" tags become "export symbols" tags
-  copts=["/DOPENTRACING_EXPORTS"],
+cc_import(
+  name = "opentracing_so_import",
+  shared_library = ":libopentracing.so",
 )
 
+## SHARED ##
+
 cc_library(
-    name = "opentracing_unix",
-    srcs = glob(["src/**/*.cpp"], exclude=["src/dynamic_load_unsupported.cpp", "src/dynamic_load_windows.cpp"]),
-    hdrs = glob(["include/opentracing/**/*.h"]) + [
-        ":include/opentracing/config.h",
-        ":include/opentracing/version.h",
-    ],
+    name = "opentracing",
+    visibility = ["//visibility:public"], # this is the only one that needs to be public
+    hdrs = hdrs,
     strip_include_prefix = "include",
     deps = [
       "//3rd_party:expected",
       "//3rd_party:variant",
-    ],
-    linkopts = [
-      "-ldl",
-    ],
+    ] + select({
+        "@bazel_tools//src/conditions:windows": [":opentracing_dll_import"],
+        "//conditions:default": [":opentracing_so_import"],
+    }),
+    # linkopts = [
+    #   "-ldl",
+    # ],
```

Comment:
@rnburn I'm not sure what the purpose of this linker flag is and haven't been able to find docs for it online.

Reply:
dl is the library for the dlopen and related system calls. See http://man7.org/linux/man-pages/man3/dlopen.3.html